k2/csrc/arcsort.h

```diff
@@ -0,0 +1,57 @@
+// k2/csrc/arcsort.h
+
+// Copyright (c)  2020  Xiaomi Corporation (authors: Daniel Povey
+//                                                   Haowen Qiu)
+
+// See ../../LICENSE for clarification regarding multiple authors
+
+#ifndef K2_CSRC_ARCSORT_H_
+#define K2_CSRC_ARCSORT_H_
+
+#include <vector>
+
+#include "glog/logging.h"
+#include "k2/csrc/fsa.h"
+
+namespace k2 {
+/**
+    Sort arcs leaving each state in the input FSA on label first and
+    then on `dest_state`
+ */
+class ArcSorter {
+ public:
+  /* Lightweight constructor that just keeps const references to the input
+     parameters.
+     @param [in] fsa_in  The input FSA
+  */
+  ArcSorter(const Fsa &fsa_in) : fsa_in_(fsa_in) {}
+
+  /*
+    Do enough work that know now much memory will be needed, and output
+    that information
+        @param [out] fsa_size   The num-states and num-arcs of the output FSA
+                                will be written to here
+  */
+  void GetSizes(Array2Size<int32_t> *fsa_size);
+
+  /*
+    Finish the operation and output the arc-sorted FSA to `fsa_out` and
+    arc mapping information to `arc_map` (if provided).
+    @param [out] fsa_out   The output FSA; Will be arc-sorted;
+                           Must be initialized; search for 'initialized
+                           definition' in class Array2 in array.h for meaning.
+    @param [out] arc_map   If non-NULL, will output a map from the arc-index
+                           in `fsa_out` to the corresponding arc-index in
+                           `fsa_in`.
+                           If non-NULL, Must be allocated with size num-arcs
```

Comment:
For `num-arcs of fsa_out`, I prefer if you make this more explicit by writing something usable as code, e.g. fsa_out->size or whatever it happens to be.

Reply:
OK, will change those files.